### ISSUE TITLE
FormControl.setValidators...why is there no getValidators?

### ISSUE THREAD

**Issue by Icepick** [angular/angular#13461]
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[x] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
We have an AbstractControl with basic validators like (required, maxLength) in a ReactiveForm via TypeScript FormBuilder. To those basic validators we added a custom validator which dynamically change the validation strategy based on a dropdown value. Currently we are using the `setValidators()` method inside another component (AbstractControl is included via `@Input()`). The main problem is that it overwrites existing validators.  
Example:

App Component
```
this.formBuilder.group({
      zip: ['', [Validators.required, Validators.maxLength(10)]] 
    });
```
App Template
```
<zip ...
        [control]="form.controls.zip">
</zip>
```
ZIP Component
```
@Input() control: AbstractControl;

this.control.setValidators([
      // redundant to formBuilder
      Validators.required,
      Validators.maxLength(10),
      // custom validation based on a dropdown value (via. valueChange Detection)
      validateZipFn(countryCode)]
    );
```

**Expected behavior**
To stay flexible we don't want to overwrite all validators. The below code would illustrate the behavior with a `getValidators()` method.

App Component
```
this.formBuilder.group({
      zip: ['', [Validators.required, Validators.maxLength(10)]]
    });
```
App Template
```
<zip ...
        [control]="form.controls.zip">
</zip>
```
ZIP Component
```
@Input() control: AbstractControl;

let listOfAllValidationRules = this.control.getValidators().push(validateZipFn(countryCode)]);
this.control.setValidators(listOfAllValidationRules);
```

**What is the motivation / use case for changing the behavior?**
Being more flexible with dynamic validation.

* **Angular version:** ~2.1.2

* **Browser:** all
 
* **Language:** TypeScript / ES6

**Comment by DzmitryShylovich** [comment]
Duplicate of https://github.com/angular/angular/issues/10567?

**Comment by DzmitryShylovich** [comment]
@Icepick there's nothing to return. validators are stored as a single object, not an array.

**Comment by Bidthedog** [comment]
@DzmitryShylovich Does that mean this cannot be implemented? Personally, I only need a readonly copy of the validators with the metadata attached.

**Comment by xeii** [comment]
I am building a directive where and object of validation messages is passed in. This is simply an object with the validation rule name is they key and the value is the message to be displayed. I would like to compare this object and the set validators of the NgControl, so that the user is warned if a certain message is missing.

**Comment by oehm-smith** [comment]
I would like to see this also for similar reasons to @Bidthedog.  For example, if a `Validator.requried` is on a control then we'd like to programatically indicate this on the form.  Read-only is all we require.  Currently the best I can do is define an array on the component and pass that to the form control definition.  Its going to get ugly pretty quickly!

**Comment by Bidthedog** [comment]
Exactly. I basically have three big nasty duplicate definitions in my form service now. The below code snippets are a tiny excerpt from a rather large, 6 page form I'm putting together. This is for page one (details) and the start of page 2 (medical):

**To create / define:**
```
  private createForm() {
    this.form = this.fb.group({
      details: this.fb.group({
        NHSNumber: [null, [Validators.required, Validators.pattern(/^\d{3}-\d{3}-\d{4}$/)]],
        titleId: [null, [Validators.required]],
        firstName: [null, [Validators.required, Validators.maxLength(10)]],
        middleNames: null,
        lastName: [null, [Validators.required]],
        DOB: [null, [Validators.required]],
        genderId: [null, [Validators.required]],
        regionId: [null, [Validators.required]],
        maritalStatusId: [null, [Validators.required]],
        occupationTypeId: [null],
        occupationId: null,
        telephoneNo: null,
        mobileNo: null,
        houseNumber: null,
        streetName: null,
        town: null,
        city: null,
        county: null,
        postcode: null,
        GPId: null,
        nextOfKinId: null,
        livesWith: null
        //vehicleLicenseTypeId?: number[]
      }),
      medical: this.fb.group({
        gpId: [null, [Validators.required]]
      })
    });
  }
```
**To populate:**
```
  private populateForm() {
    let m = this.model;
    this.form.setValue({
      details: {
        NHSNumber: this.getTextValue(m.NHSNumber),
        titleId: this.getLookupValue(m.titleId),
        firstName: this.getTextValue(m.firstName),
        middleNames: this.getTextValue(m.middleNames),
        lastName: this.getTextValue(m.lastName),
        DOB: this.getDateValue(m.DOB),
        genderId: this.getLookupValue(m.genderId),
        regionId: this.getLookupValue(m.regionId),
        maritalStatusId: this.getLookupValue(m.maritalStatusId),
        occupationTypeId: this.getLookupValue(m.occupationTypeId),
        occupationId: this.getLookupValue(m.occupationId),
        telephoneNo: this.getTextValue(m.telephoneNo),
        mobileNo: this.getTextValue(m.mobileNo),
        houseNumber: this.getTextValue(m.houseNumber),
        streetName: this.getTextValue(m.streetName),
        town: this.getTextValue(m.town),
        city: this.getTextValue(m.city),
        county: this.getTextValue(m.county),
        postcode: this.getTextValue(m.postcode),
        GPId: this.getLookupValue(m.GPId),
        nextOfKinId: this.getLookupValue(m.nextOfKinId),
        livesWith: this.getTextValue(m.livesWith)
      },
      medical: {
        gpId: this.getLookupValue(m.GPId) || 1
      }
    });
```

Validation rules. Not complete, but you can see how this is going to get messy, fast

```
  private formValidationMessages = {
    details: {
      NHSNumber: {
        required: '\'NHS Number\' is required',
        pattern: '\'NHS Number\' must be in the format \'NNN-NNN-NNNN\' where N is a digit'
      },
      titleId: {
        required: '\'Title\' is required'
      },
      firstName: {
        required: '\'First Name\' is required'
      }
    },
    medical: {
    }
  };
```

I started working on defining all the form rules in one object set, which I'm thinking of iterating through / projecting into the functions listed above (though I've not got that far yet):

```
  private formValidationMessages = {
    details: {
      NHSNumber: {
        required: '\'NHS Number\' is required',
        pattern: '\'NHS Number\' must be in the format \'NNN-NNN-NNNN\' where N is a digit'
      },
      titleId: {
        required: '\'Title\' is required'
      },
      firstName: {
        required: '\'First Name\' is required'
      }
    },
    medical: {
    }
  };
```

**Comment by JamesHenry** [comment]
I also just got hit by not being able to retrieve the configured Validators for a FormControl.

I worked around it by adding a utility service which wraps the config object before it is passed into the FormBuilder, and keeps track of a mapping between control names and their given Validators.

It would naturally be great to remove the need for this extra service 😄

**Comment by Bidthedog** [comment]
Well, since I posted here I've consolidated all my form configuration into one big object, and wrote a few helper methods that do the mapping between form and model based on the config. It all seems way too complicated, but it could just be the learning curve I've had.

**Comment by RabbitHunter99** [comment]
Hi,
just wondering if there is any word on this feature. It seems to me that being able to visually mark certain items as required is a basic feature expected of any modern forms UI. Yet since there is no way to retrieve the information whether a field is required or not, it is impossible to implement this feature using reactive forms.

**Comment by Bidthedog** [comment]
It's not impossible, it's just a LOT of work.

**Comment by Toub** [comment]
@Bidthegod do you mean it would be a LOT of work to implement this feature in angular, or it would be a LOT of work to retrieve the validators in an angular component.

If it is possible to retrieve the validators in a component, any input about how to do is welcome!

Or if it is easier for certain component types (input, select), it can be useful too.

**Comment by Bidthedog** [comment]
@Toub What I ended up doing was building a "config" object structure that contained every individual form field, and all its settings (display, validators and rules etc). I then wrote a few helper methods to query the config, which builds up the FormGroup. Each form has a form service, and inherits from a base component.

I wouldn't say it's an easy fix, and it's tons of code, but it ended up being fairly elegant and easy to create a fairly complex form. Threw some of the code in a gist for you to look at, but I've not got time to modularise properly it at the moment. There are quite a few files missing from this - including all the custom form input components I wrote.

https://gist.github.com/Bidthedog/1dc7d10cda1759061c09f44f7b48cbf3

**Comment by Toub** [comment]
@Bidthedog it doesn't feet my needs but thanks for your answer, as the gist can be useful for others.

**Comment by rodolfojnn** [comment]
Hey guys, I created a simple helper to return the validators (just for the required, but can be extended for others):

```
getValidators(_f) {
    return Object.keys(_f).reduce((a, b) => {
      const v = _f[b][1];
      if (v && (v === Validators.required || v.indexOf(Validators.required) > -1)) {
        if (!a[b]) { a[b] = {}; }
        a[b]['required'] = true;
      }
      return a;
    }, {});
  }

const _f = {
      id: [],
      name: [null, Validators.required],
      phone: [],
      email: [null, [Validators.required, Validators.email]]
    };
    this.frmMain = this._fb.group(_f);
    console.log(this.getValidators(_f));    // {name: {"required": true}, email: {"required": true}}
```

Fits my needs 👍

**Comment by erikrasmussen23** [comment]
I agree  having a readonly list of validators on current control would be very helpful.

**Comment by herinruhela** [comment]
+1 for this feature

**Comment by rajucp** [comment]
I managed to do it in a different way as below. I used primarily for finding mandatory fields. But You can add as many properties as you would want such as add all validators there too as below.
**1. Define an interface**
```
export interface StandardAbstractControl extends AbstractControl {
  mandatory: boolean;
  validators: ValidatorFn[]
}
```
**2. And add a dummy validator function which returns null so no errors.**
```
 someService.getMandatorySetterValidator(isRequired: boolean, listOfValidators: ValidatorFn[]): ValidatorFn {
    return (c: StandardAbstractControl): ValidationErrors => {
      c.mandatory = isRequired;
      c.validators = listOfValidators;
      return null;
    };
  }
```

add to your formcontrol
```
formcontrol.setValidators([
            this.someService.getMandatorySetterValidator(true, real-validators-list),
            real-validators-list
          ]);
```
now you can use it anywhere you want.

**Comment by darkolakovic** [comment]
+1 for read-only list

**Comment by jfox459** [comment]
+1

**Comment by RidelHI** [comment]
I need that functionality to dynamically add validators without losing those that already exist. Something like:
`Object.keys(this.myForm.controls).forEach(key => {
  if (map.get(key)) {
    this.myForm.get(key).setValidators(this.myForm.get(key).getValidators().push(Validators.required))
  }
});`
Fully explained in: [https://stackoverflow.com/questions/46852063/how-to-dynamically-add-validators-to-the-forms-in-angular-2-4-using-formbuilde](https://stackoverflow.com/questions/46852063/how-to-dynamically-add-validators-to-the-forms-in-angular-2-4-using-formbuilde).
This must be something common

**Comment by ppham27** [comment]
It's actually not too hard to dynamically add validators with composeValidators (https://github.com/angular/angular/blob/master/packages/forms/src/directives/shared.ts#L139). I think removing validators is the bigger issue. You can do it with a boolean but then your validator is no longer a pure function.

**Comment by mtinner** [comment]
this function should work wor FormGroups and FormControls to determine required Validators
  
```
      export const hasRequiredField = (abstractControl: AbstractControl): boolean => {
    	if (abstractControl.validator) {
    		const validator = abstractControl.validator({}as AbstractControl);
    		if (validator && validator.required) {
    			return true;
    		}
    	}
    	if (abstractControl['controls']) {
    		for (const controlName in abstractControl['controls']) {
    			if (abstractControl['controls'][controlName]) {
    				if (hasRequiredField(abstractControl['controls'][controlName])) {
    					return true;
    				}
    			}
    		}
    	}
    	return false;
    };
```

**Comment by strizzaflex** [comment]
@mtinner That was a really great find. That is getting me the initial set. I think it would still be worth adding to the API to allow for a return of an observable of the validators. That way finding a conditional required control would be very simple. I'm not sure the proposed workaround here can be applied to catch those instances.

**Comment by paolo-galfione** [comment]
In a very nice presentation of Angular Connect 2017 @kara ends by saying "custom forms controls are awesome & not scary". Good, but this important feature for custom controls is missing

**Comment by danderwald** [comment]
+1

**Comment by seedy** [comment]
No time for a PR but here's what I came up with : 
```
import {FormControl} from '@angular/forms';
import {AsyncValidatorFn, ValidatorFn} from '@angular/forms/src/directives/validators';
import {Observable} from 'rxjs/Observable';
import {Subject} from 'rxjs/Subject';
import {AbstractControlOptions} from '@angular/forms/src/model';

export class DescendingFormControl extends FormControl {

    private asyncValidationChangesSubject = new Subject<AsyncValidatorFn | AsyncValidatorFn[]>();
    private validationChangesSubject = new Subject<ValidatorFn | ValidatorFn[] | null>();

    public readonly validationChanges: Observable<ValidatorFn | ValidatorFn[] | null>;
    public readonly asyncValidationChanges: Observable<AsyncValidatorFn | AsyncValidatorFn[]>;

    constructor(formState?: any, validatorOrOpts?: ValidatorFn | ValidatorFn[] | AbstractControlOptions | null, asyncValidator?: AsyncValidatorFn | AsyncValidatorFn[] | null) {
        super(formState, validatorOrOpts, asyncValidator);
        this.validationChanges = this.validationChangesSubject.asObservable();
        this.asyncValidationChanges = this.asyncValidationChangesSubject.asObservable();
    }

    public setValidators(newValidator: ValidatorFn | ValidatorFn[] | null): void {
        super.setValidators(newValidator);
        this.validationChangesSubject.next(newValidator);
    }

    public setAsyncValidators(newValidator: AsyncValidatorFn | AsyncValidatorFn[]): void {
        super.setAsyncValidators(newValidator);
        this.asyncValidationChangesSubject.next(newValidator);
    }
}

```

If anyone's up to do the whole PR job before I get some free time, go for it!

**Comment by pete4dev** [comment]
+1

**Comment by mihaimascas** [comment]
+1

**Comment by Lowkey2224** [comment]
+1

**Comment by urechm** [comment]
The following snippet adds the Validators.required validator to the existing validators on the issueCountryControl, which solves the initial problem of the missing setValidators function.
```
const issueCountryControl = this.formGroup.get('issueCountry');
issueCountryControl.setValidators(Validators.compose(
   [issueCountryControl.validator, Validators.required])
);
```

**Comment by jeremyben** [comment]
### The problem : It's trapped
An array of present validators exists but is not available to us...
Have a look at the Validators static method `compose`, which is the method used behind the scenes when you define an array of Validators on your form control : 
```typescript
 static compose(validators: (ValidatorFn|null|undefined)[]|null): ValidatorFn|null {
    if (!validators) return null;

    // Here it is
    const presentValidators: ValidatorFn[] = validators.filter(isPresent) as any; 

    if (presentValidators.length == 0) return null;

    return function(control: AbstractControl) {
      return _mergeErrors(_executeValidators(control, presentValidators));
    };
  }
```
> https://github.com/angular/angular/blob/master/packages/forms/src/validators.ts#L342

As you can see, `presentValidators` is what we're looking for, but it's _trapped_ in a closure, so inaccessible, and only value you get from `compose` is an object with all the triggered errors for a given value.

So the only way to ~~get~~ infer the validators is to manually trigger the errors on your form control, which will reference the validators names.

### The hack : Trigger an error on a different control
You don't want to trigger an error on your binded form control. You should instead create a new form control, free from any binding to your template, and apply the same Validator functions to him, then trigger the errors you want by setting a value : 
```typescript
// Your form control, actually binded to a form in your template
emailControl = new FormControl('', [Validators.required, Validators.email])

// Elsewhere in your code
// You use dummy form controls, on which you apply your original form control validators and various values, to trigger the errors.

const errorsWhenEmpty = new FormControl('', this.emailControl.validator, this.emailControl.asyncValidator).errors
const errorsWhenOneChar = new FormControl('x', this.emailControl.validator, this.emailControl.asyncValidator).errors
```
It is far from ideal, but it's the least dirty way I have found, until we can get our hands on the validators themselves.

**Comment by GGGErnest** [comment]
I will suggest just to extend from the FormControl class and save in the fields of the children class a copy of the Validator functions passed into the constructor, something like this:
`export class ExtFormControl extends FormControl {
  private _syncValidators: ValidatorFn | ValidatorFn[];

  private _asyncValidators: AsyncValidatorFn | AsyncValidatorFn[];

  constructor(formState?: any, validatorOrOpts?: ValidatorFn | ValidatorFn[] | null,
              asyncValidators?: AsyncValidatorFn | AsyncValidatorFn[] | null) {
    super(formState, validatorOrOpts, asyncValidators);
    this._syncValidators = validatorOrOpts;
    this._asyncValidators = asyncValidators;
  }


  getSyncValidators(): ValidatorFn[] {
    if (typeof this._syncValidators === 'function') {
      return [this._syncValidators];
    } else {
      return this._syncValidators;
    }
  }

  getAsyncValidators(): AsyncValidatorFn[] {
    if (typeof this._asyncValidators === 'function') {
      return [this._asyncValidators];
    } else {
      return this._asyncValidators;
    }
  }

  setValidators(newValidator: ValidatorFn | ValidatorFn[] | null): void {
    super.setValidators(newValidator);
    this._syncValidators = newValidator;
  }

  setAsyncValidators(newValidator: AsyncValidatorFn | AsyncValidatorFn[] | null): void {
    super.setAsyncValidators(newValidator);
    this._asyncValidators = newValidator;
  }

}`

**Comment by moniuch** [comment]
I am all for a read-only list of validators. There are cases where, using ReactiveForms, one would expect to configure constraints like maxLength, via FormBuilder code rather than template. But it is currently necessary to place maxlength in HTML, in order to be able to define some behaviors (like moving focus to next field, a typical case in credit card inputs), or automatically marking labels as required, etc.

**Comment by taylorhutchison** [comment]
I have a need to programmatically set validators that update as the form gets filled in. It would be really nice to be able to read the validators on a FormControl. I agree with @moniuch that using ReactiveForms one would expect to be able to do this. Just voicing my hope it gets implemented.

**Comment by Mallfurion** [comment]
any updates, folks?

**Comment by BenRacicot** [comment]
Hmm I'm seeing this compound an interesting issue in [19851 ](https://github.com/angular/angular/issues/19851)

**Comment by marmelin** [comment]
I solved it for me with check the "validator" property of the control. It gives back the object with **ACTIVE** validators. If validator's are set and if the condition for the validator to be true is fulfilled. 
In my example it is even a **REACTIVE FORM** used in a **NESTED COMPONENT**.

## Parent Component
#### The validators in ReactiveFormGroup in PARENT component
```javascript
return this.formBuilder.group({
      username: ['', [Validators.required, Validators.min(3), Validators.email]],
     ...
});
```
The names are only for testing purposes
```html
<app-input name="txtfield" 
[parentForm]="loginForm" parentFormControlName="txtfield" 
></app-input>
```

## Nested Component
```html
 <div [formGroup]="parentForm" class="form-control-group">
   ...
        <input #inputField  name="{{parentFormControlName}}" 
            formControlName="{{parentFormControlName}}" 
            matInput 
            placeholder="{{placeholderTxt}}"
        >
  ...
    </div>
```
In nested component I check only once after AfterViewIinit, because I only need "required" validator. 
```javascript
ngAfterViewInit() {
    // Lookup if this field is required or not
    if (this.parentForm.controls[this.parentFormControlName].validator != null) {
    // MAYBE CHECK FOR required field in returned OBJECT!!
      console.log('Validators of ' + this.parentFormControlName + ' = ', 
           JSON.stringify(
             this.parentForm.controls[this.parentFormControlName]
                      .validator(this.parentForm.controls['username']);)
            );

    } else {
      console.log('No validator set for component ' + this.parentFormControlName); 
    }
```
Output in browser console while debugger stopped:
```javascript
[DEBUG] STATUS of username =  {"required":true}
>this.parentForm.controls[this.parentFormControlName].validator(this.parentForm.controls[this.parentFormControlName])
{required: true}
required: true
__proto__: Object

```
Otherwise - means if no validator is set - it shows "null"!

### Other situations
#### If input text field has a default value
If the input-text-field has a default value, this would **NOT** show the required value to be true, but I only wanna highlight the field (= line bold) if the field is required and not filled. 
#### Other value than required
You can always check any other validator status, if you watch changes with maybe "DoCheck" or "OnChanges" to get maybe "email" or "min" validators.... if you need this.

**Comment by Vinutha-BK** [comment]
@Angular team , it's really necessary to have the API to check current validator list and remove specific validator at run time. This solve many issues.

**Comment by marmelin** [comment]
Why "control.errors" and "controls.validators" doing preaty mutch the same thing?
".**errors**" should give all errors than an error state is reached.
".**validators**" should give back all registered validators, for this control. 
As simple as that?

**Comment by dominique-mueller** [comment]
I'd very much like to see this feature as well. But! One of the big issues I'm seeing is that validators are (usually) pure functions - meaning they can be nested, even deeply.

In our projects we have lots of domain-related validators which, then, are constructed out of other validators (and so on). Our IBAN validator, for instance, uses a `exactLength`, a `pattern` and a IBAN check digit validator under the hood. Following this pattern leads to to more specific validation errors, and allows us to use error messages which describe the issue in greater detail (aka better for Usability).

So, when considering a new method for retrieving all validators, the questions is: Will only be the "top validators" be returned? Plus, what's the name of those validators? Overall, it seems to me that, with the current implementation of validation, there is no easy solution ...

**Comment by ganqqwerty** [comment]
something like addValidators would also be useful

**Comment by ganqqwerty** [comment]
So folks, the relevant source code is this: 
[packages/forms/src/model.ts](https://github.com/angular/angular/blob/master/packages/forms/src/model.ts)
[packages/forms/src/validators.ts](https://github.com/angular/angular/blob/master/packages/forms/src/validators.ts)
and maybe [packages/forms/src/directives/validators.ts](https://github.com/angular/angular/blob/master/packages/forms/src/directives/validators.ts)

I found the following problems: 
* first of all the validator is just a stupid function object with no way of identifying it. Even if you could get an array of functions, what would you do then? We need some way to identify that this one is a regex validator, and this one is a required validator if we want to add them or remove them individually. 
* secondly, when you setValidators, [the compose function is eventually called](https://github.com/angular/angular/blob/master/packages/forms/src/validators.ts#L344). It lumps all validator function together into a single function and actually calls them [by calling internal _executeValidators function](https://github.com/angular/angular/blob/master/packages/forms/src/validators.ts#L386). So in the end we don't have the original array of validators, we only have a huge mega-validator with the list of all errors merged.

**Comment by mlc-mlapis** [comment]
@ganqqwerty … Correct. That's the reason why `addValidator(...)` or `removeValidator(...)` doesn't exist. There is no identification. The validator itself doesn't know on which control is running.

**Comment by ganqqwerty** [comment]
Thanks for clarifying this to me! Does it mean that such feature can be scheduled for implementation?

**Comment by andreElrico** [comment]
Without this feature it impossible to **remove** a **specific validator** from a **specific control**.

This is what I need right now and I cant solve this. Its crasy how sometimes super basic stuff is just not possible to do with angular.

`this.myAbstractFormControl.removeValidatorAndKeepTheRestAsIs(Validators.required)`

**Comment by lazarljubenovic** [comment]
@andreElrico What's your use-case for removing a validator? The more I think about this, the less I see the point in ever changing validators on a form. If you need to change validators dynamically based on a different field in the form, then your placement of the validator is wrong: what you want is a form group validator, not form control validator.

**Comment by andreElrico** [comment]
Im using the same formGroup that is provided via e.g. input dynamically in two senarios. One where I CAN have required on each field and the second case where I remove the required IF there and apply a atLeastOneField validator to the form.

Here is my [stackblitz](https://stackblitz.com/edit/angular-wr2nhd) Im working on.

I dont understand why some protect the design flaw that the angular team has produced.
Angular Form-Validation should be [CRUD](https://en.wikipedia.org/wiki/Create,_read,_update_and_delete).

**Comment by lazarljubenovic** [comment]
@andreElrico As I said, if your form has two inputs that have validations depndent on each other, then your validator should be in the form group, not in the form contol.

For example, if you have two password fields, you shouldn't change the validator of the second control each time the first control is change. Instead, the validator should go on a group which contains two password fields. Each password field has validation that password must be at least 8 characters long, but the group has the validator that they need to be the same. On its own, each password field isn't aware of existance of the other one. The group holds them together.

If you have something like

```
( ) a
( ) b
( ) other [ _____ ]
```

where if user chooses "other", the input field is required next to it, then, again, the validator "required" shouldn't be toggled on the input field. Instead, the whole form group needs validation: the radio button control needs to be one of "a", "b" or" other", and if "other", then input field must be present.

You shouldn't have a listener on radio button and then toggle input field's "required" validator based on the value of the radio buttons.

I don't see why you believe validation should have a CRUD API. I'm not defending anything, I'm speaking reasonable and with arguments. I'm not saying methods for getting, removing and toggling validators should't be implemented, I'm saying that almost everyone here has a use-case for CRUD validations which are actually anti-patterns.  A control is a standalone unit and it doesn't change its validation rules during its lifetime. When controls are dependent, then they are put in a group and the group is validated.

**Comment by andreElrico** [comment]
thank you @lazarljubenovic for your comments and time.
I feel you havent looked closely at my stackblitz. Im applying the "wrapping" logic in the formGroup for one scenario. I just have two distinct scenarios for my form that need the validators to change dynamically. And at that "stage" where I deal with the formGroup my validators are already "composed", so no CRUD for me :D.

I can work around this problem and hope this gets fixed someday. 
167 👍 seem to reflect a high desire for this feature.

If you have time, I would love to find out more:

"A control is a standalone unit and it doesn't change its validation rules during its lifetime." do you have any resources on that?

"When controls are dependent, then they are put in a group and the group is validated." This acually is has its limits as well.
+ https://github.com/angular/angular/issues/31105

Also not related to this ISSUE:
If the group takes over the validation of the controls the errors will be set to the group and therefore not be reflected visually on the controls. So we have to set and remove the errors manually to the child controls. (Thats a big pain in the ***).

I wish all a good weekend.

**Comment by Da13Harris** [comment]
I certainly understand @lazarljubenovic's perspective. However, I work on highly dynamic, multi-part, enterprise forms that require complex composition... lots of hiding/showing, validation requirement dependency chains, etc. The way we handle this complexity is to subscribe to valueChanges on any form control, group, or array that is responsible for triggering dependencies, then using setValidators or clearValidators for each scenario. Since there's no way to target specific validators, we simple store each set of validation rules that might apply at any given time and reapply them. It sure would be nice to target specific validators for removal, which seems like it could still be architected with immutability in mind.

**Comment by lazarljubenovic** [comment]
>> "A control is a standalone unit and it doesn't change its validation rules during its lifetime."
>
> do you have any resources on that?

Nope; it's really just my interpretation of the design of forms.

> If the group takes over the validation of the controls the errors will be set to the group and therefore not be reflected visually on the controls. 

To be honest, it's dubious even in terms of UI where those errors are supposed to go. When two fields don't play together, but are far apart in the form, where _should_ you write the error? It seems to me like you should take a step back and fix the UX of the form firstly.

Of course, forms are a wild beast and they cannot fit every use-case. When I have a VERY complex form with _really_ crazy controls, I don't use `@angular/forms` at all. I see `@angular/forms` as a great utility which covers 95% of use-cases, and allows me to quickly build a form without much effort. 

Just like I like using `Array#map`, `Array#filter` and `Array#reduce`, sometimes a good old `for` loop with an index variable is the best solution for a more complex iteration. 

I'd just like to make it clear again that I'm not _against_ adding these methods; I just feel like the _majority_ of people would misuse them. Maybe it would be a footgun. Then again, I don't think that a useful shouldn't be added to a framework just because someone can misuse it. You can also brush your teeth with toilet water and we don't ban toilets because of that.

**Comment by ganqqwerty** [comment]
> When I have a VERY complex form with really crazy controls, I don't use @angular/forms at all. I see @angular/forms as a great utility which covers 95% of use-cases, and allows me to quickly build a form without much effort.

Is this the official market positioning of the Angular Forms? I thought of it more as of all-purpose tool that is useful for huge projects.

**Comment by Bidthedog** [comment]
> 
> 
> > When I have a VERY complex form with really crazy controls, I don't use @angular/forms at all. I see @angular/forms as a great utility which covers 95% of use-cases, and allows me to quickly build a form without much effort.
> 
> Is this the official market positioning of the Angular Forms? I thought of it more as of all-purpose tool that is useful for huge projects.

"Reactive Forms" are meant to be for the complicated, highly customised use-cases... such as being able to completely control your validation ;)

**Comment by lazarljubenovic** [comment]
You can completely control it. You just need to properly determine WHAT you're validating: a group or a control.

**Comment by joeylgutierrez** [comment]
has this made any traction? I would really love to dynamically change my validators with only a formGroup as an input to keep everything decoupled!!! Reactive forms are awesome but this is my only hold up.

**Comment by mswilson4040** [comment]
Sooooooooo how am I supposed to unit test my code that dynamically adds validators?

**Comment by bwronin** [comment]
I'm a big fan of Reactive Forms and love what I can do with them, but small things like not being able to ask the FormControl about its validators seems like an oversight.  Given that there are so many state properties on a FormControl, what prevents at a minimum making a `.required` property available?  Imagine if this was the case in everyday life...

Me: Hi, I'm here for my appointment.
Receptionist: Sorry we can't see you.
Me: Why not? I have an appointment now.
Receptionist: *Shrugs* I don't know. 

I'm generalizing a bit, but if I'm handing the FormControl property values I should also be able to access those values as well.

**Comment by lazarljubenovic** [comment]
> what prevents at a minimum making a .required property available

The fact that you can create your own `required` validator. Angular cannot know what your function does.

**Comment by bwronin** [comment]
100% get that, I'm not saying it be based on a function.  Knowing that a field is required is just as basic as knowing its enabled/disabled which is configurable, accessible, and reflected in the UI. 
Is there anything that can be done so that developers can get similar functionality for a field being required so they can do things like conditionally show required markers just by working with the FormControl?

**Comment by Totati** [comment]
Hi, this is our latest hack, to automatically mark inputs required: https://stackblitz.com/edit/material-input-auto-required-asterisk. Same concept can apply to mat-select, or any other library.

**Comment by lazarljubenovic** [comment]
> 100% get that, I'm not saying it be based on a function. Knowing that a field is required is just as basic as knowing its enabled/disabled which is configurable, accessible, and reflected in the UI.

I'm not so sure. “Requiredness” is very specific to the applications' business logic. Does an empty string pass a "required" condition? What about a zero? An empty array for a multiple select? False for a checkbox? Null, undefined?

> Is there anything that can be done so that developers can get similar functionality for a field being required so they can do things like conditionally show required markers just by working with the FormControl?

You can already conditionally show/hide errors based on any condition you like. The presence of the error doesn't necessarily mean you have to display it anywhere. I don't see how querying for validators instead of simply querying for errors (which is available) would help you here.

**Comment by bwronin** [comment]
> I'm not so sure. “Requiredness” is very specific to the applications' business logic. Does an empty string pass a "required" condition? What about a zero? An empty array for a multiple select? False for a checkbox? Null, undefined?

I could argue that a field being enabled/disabled is also very specific to an application' business logic too but the FormControl provides functionality for a developer to manage that via the constructor and accessors.  Being required is a fuzzy line and I get where you're coming from.  This may be personal preference, but it's as much a presentational state (like enabled/disabled, valid/invalid) as it is an indication of the value's validity. 

> You can already conditionally show/hide errors based on any condition you like. The presence of the error doesn't necessarily mean you have to display it anywhere. I don't see how querying for validators instead of simply querying for errors (which is available) would help you here.

Yes I can conditionally hide/show errors, but once those errors are resolved any presentational pieces that hinge off of the expected state are no longer available.  I can't show an asterisk based on the validators because once those validators are satisfied they are no longer available. 
I guess for now I'll just have to work around it.

**Comment by lazarljubenovic** [comment]
> a field being enabled/disabled is also very specific to an application' business logic too

I can't see how.

> I can't show an asterisk based on the validators

So instead add a validator based on your own "required" semantics. Not everything can be a one-liner.

**Comment by bwronin** [comment]
I'm looking for any reasonable way to support the scenario, not necessarily a one-liner.  I'm also looking to understand your perspective in case mine is misaligned.    

What if my semantics were just vanilla and only needed to leverage the Validators.required validation?  In the StackBlitz example below, if I add it to a text field and enter a value, the only way to know about it is via the `field.errors` object and look for the property associated with the Validators.  However, when the Validator is satisfied, that property no longer exists.  Maybe my approach is incorrect, is there a better way to leverage the FormControl to achieve this that I'm not taking advantage of?

[StackBlitz example](https://stackblitz.com/edit/angular-d8eio7)

**Comment by lazarljubenovic** [comment]
You're not supposed to toggle validators based on validity of a different field. Create a group validator and check the value of the object containing both the name and secondary name. See [this comment](https://github.com/angular/angular/issues/13461#issuecomment-533399094) for more info.

As for the star... Well, your approach is not "incorrect" in a sense "that's a terrible idea", but it's an idea which won't work with the way validators are currently envisioned. So I'd rather say your approach is _incompatibile_ with how forms are envisioned to be used right now.

You'll need to create your own mini form creation library if you want to quickly achieve the effect you describe. Here's a quick sketch of how I'd approach the problem if I was faced with this problem. For example, you can have a function `createForm` which accepts a `spec` object like this:

```
spec = {
  name: {
    initialValue: '',
    isRequired: (formValue) => true, // always required
  },
  secondName: {
    initialValue: '',
    isRequired: (formValue) => isFilledIn(formValue.name), // required only if name is filled in
  }
```

`isFilledIn` is your function which determined whether a value is filled in or not. Probably something like `val => val != null && value != ''`.

This `createForm` would return the same form as if you've created it like this:

```
this.fb.group({
  name: [''],
  secondName: [''],
}, {
  validators: [
    (group) => {
      const value = group.value
      const errors = {}
      if (spec.name.isRequired(value) && !isFilled(value.name)) {
        errors['name-required'] = true
      }
      if (spec.secondName.isRequired(value) && !isFilled(value.secondName)) {
        errors['secondName-required'] = true
      }
      return errors // or null if no errors found
    }
  ]
})
```

The errors will now be on the group itself. The error for `secondName` indeed belongs there -- you cannot validate `secondName` on its own, since its validity depends on a different field (so you validate the group of those fields instead). The error for `name` should belong to `name`, but this sketch of implementation would put it on the form group. This might or might not work for you. You could introduce a field on the `spec` object which would further tweak the configuration like this.

Now in the template you use your form as usually, since you have the `@angular/forms` object at hand. You'll need to print some errors from the form group and some from the controls (or maybe all from the group if that works for you). The condition for toggling the star would be `spec[key].isRequired(form.value)`.

The possibilities if how precisely you're going to implement this are endless, and it all depends on such tiny decisions that I don't think any forms API would satisfy everyone. So depending on what you expect of a form in your own project, in my opinion it's best to introduce a “semantic” wrapper around the form builder, which will do additional stuff based on your requirements.

**Comment by bwronin** [comment]
Thanks for the approach and for spending the time referencing the previous comment, the thread history is quite long :) 

I think that's where part of my issue has been; properly managing field dependencies between controls that are grouped and ones that live outside of the group.  The spec approach has been my fallback in most areas so I'm glad it seems to be how you'd approach it as well.

> that I don't think any forms API would satisfy everyone
True, not sure any provided API meets that sort of coverage!

**Comment by jalu14** [comment]
Any update? almost 4 years since the creation of this issue.
We are currently creating custom utility functions to store validations, functions that derive too much from the 'vanilla' angular just to getValidators of a formControl.

**Comment by AnnaRainier** [comment]
would really need this feature

**Comment by CD-UNCC** [comment]
This would be very helpful for those of use who wrap reactive forms

**Comment by h-arora** [comment]
Very valuable thread. However, I am a little confused as to what purpose it solves by not giving access to validators on a form control. I bumped into this requirement where we have our custom component that needs to show an "*" if it is required and maybe a relevant error message right below it (if the user wants that).  The other alternate is adding to the API of the control itself by making the user set a property from the template as well. But that should not be needed as in a reactive form the user already provided his required validator (or for that matter any other validator). My custom component should be able to know that such a validator is set and it could then show/hide elements based on that. Seems a very common requirement.

**Comment by antodaniau** [comment]
A solution with a custom FormControl to have all informations about validators and potentially full control on those validators : 

```
export class CustomFormControl extends FormControl {

  public listValidator: CustomValidator[] = [];

  private constructor(formState?: any,
                      validatorOrOpts?: ValidatorFn | ValidatorFn[] | CustomCtrlOpts | null,
                      asyncValidator?: AsyncValidatorFn | AsyncValidatorFn[] | null) {

    super(formState, validatorOrOpts as ValidatorFn[], asyncValidator);

    if (validatorOrOpts instanceof CustomCtrlOpts) {
      this.listValidator = validatorOrOpts.customConf;
    }
  }

  // Builder ith default value
  public static init(formState?: any): CustomFormControl {
    return new CustomFormControl(isNil(formState) ? null : formState);
  }

  public required(): CustomFormControl {
    return this.and('required');
  }

  public email(): CustomFormControl {
    return this.and('email');
  }

  public custom(): CustomFormControl {
    return this.and('custom');
  }

  public min(min: number): CustomFormControl {
    return this.and('min', min);
  }

  public max(max: number): CustomFormControl {
    return this.and('max', max);
  }

  private and(key: TypeCustomValidator, value?: any): CustomFormControl {
    this.listValidator.push({key: key, value: value});
    this.setValidators(new CustomCtrlOpts(this.listValidator).validators as ValidatorFn[]);
    return this;
  }
}

// Add others if needed
export type TypeCustomValidator = 'min' | 'max' | 'required' | 'email' | 'custom' ;

export interface HandledValidator {
  key: TypeCustomValidator;
  function: Function;
  params?: boolean;
}

export const listHandledValidators: HandledValidator[] = [
  {key: 'required', function: Validators.required},
  {key: 'email', function: Validators.email},
  {key: 'min', function: Validators.min, params: true},
  {key: 'max', function: Validators.max, params: true},
  {key: 'custom', function: custom}
];

export interface CustomValidator {
  key: TypeCustomValidator;
  value?: any;
}

export class CustomCtrlOpts implements AbstractControlOptions {

  validators?: ValidatorFn | ValidatorFn[] | null;
  asyncValidators?: AsyncValidatorFn | AsyncValidatorFn[] | null;
  updateOn?: 'change' | 'blur' | 'submit';

  private _customConf?: CustomValidator[];

  constructor(value: CustomValidator[]) {
    this.customConf = value;
  }

  get customConf(): CustomValidator[] {
    return this._customConf;
  }

  set customConf(value: CustomValidator[]) {

    this._customConf = value;

    if (!value) {
      return;
    }

    this.validators = [];

    value.forEach(customValidator => {
      const validator = listHandledValidators.find(it => it.key === customValidator.key);
      if (validator.params) {
        (this.validators as ValidatorFn[]).push(validator.function(customValidator.value) as ValidatorFn);
      } else {
        (this.validators as ValidatorFn[]).push(validator.function as ValidatorFn);
      }
    });
  }
}

```

with formGroup initialisation like
```   
this.form = this.fb.group({
    amount: CustomFormControl.init(10).required().min(0), 
    ...
```

**Comment by tharindusathis** [comment]
+1

**Comment by stevewhitmore** [comment]
This would be a really nice feature to have. Using workarounds for this feels like a waste of time.

**Comment by tharindusathis** [comment]
> would really need this feature

Yeah this is really needed. Simply, just think if there is a way to **set** something, then there must be also a similar way to **get** it no? If you can set something easily, but for getting them back you have to spend lot of effort, then there is clearly missing something that _really needed_ .

**Comment by BenKatanHighLander** [comment]
Any progress on this issue/feature?

**Comment by mlc-mlapis** [comment]
@BenKatanHighLander  It's on the way: https://github.com/angular/angular/pull/37263#issuecomment-723333060

**Comment by mcantonbul** [comment]
This feature must be added to the roof for us to catch up with "React". Isn't there any improvement?

**Comment by lazarljubenovic** [comment]
React is lagging behind Angular, it doesn't even do anything about validation. So let them catch up, or even better yet, let's not compare apples to oranges.

**Comment by mcantonbul** [comment]
> React is lagging behind Angular, it doesn't even do anything about validation. So let them catch up, or even better yet, let's not compare apples to oranges.

Forms, as you said, is a topic that Angular is ambitious about. It is regrettable that such a simple feature has not been added for 4.5 years, despite the community's demand for it.

 By the way, the example should not be apple and orange, but apple and samsung 😄 . This is a very nice competition for software.

**Comment by lazarljubenovic** [comment]
It's a tricky one. While I understand that it's the quickest and most obvious solution to simply “get” the validator, it won't solve much. You'll add a validator `Validate.range(1, 5)`, and you'll “get” it, but you won't be able to “get” the numbers 1 and 5.

Many arguments in this thread rely on the “it's simply obvious to have it”, but it's _not_. I have several prototypes of a new Angular Forms module, but whenever I manage to do something, I either lose ergonomics of the API or make it way too complex to maintain. I'm not saying "If I can't do it, it's impossible" -- the point is, I've tried and I've seen why it's not as "simple" as everyone loves stating.

Here's a [recent comment](https://github.com/angular/angular/issues/13461#issuecomment-716081767):

> Simply, just think if there is a way to set something, then there must be also a similar way to get it no? 

No, not really. The set/get logic gets wrecked when you can "set" (actually append/add) multiple times. When you multiply a number, there's no doing back unless you know precisely what you did with it. 

Many rendering APIs work like this. You don't “set” a scale factor of 2 to blow up size of polyhedra, you multiply the existing matrix. You can't go back and you can't "get" a scale either -- there's no way to know how you arrive there. If you _do_ need this, the obvious solution is to save the matrices yourself, with your own logic in mind of how you imagined it.

There's just so many questions that need to be discussed if you want to add a “get” method. All of the solutions given in this thread introduce quite a bit of bias in the code, or have dubious practices such as using strings in plain English to distinguish between validators.

I just don't see a good solution that would be suitable for everyone -- not even for the majority.

**Comment by lazarljubenovic** [comment]
But what are we even talking about -- apparently this has been addressed: https://github.com/angular/angular/pull/39235#issuecomment-754782607

**Comment by Da13Harris** [comment]
First of all, I appreciate @lazarljubenovic's comment. I use Angular's reactive forms every day, but I certainly don't have a detailed grasp of the internal complexity. I can only try to imagine how difficult it might be to serve up a silver bullet solution that magically solves all use cases for retrieving validators.

However, I still think there are some fundamental use cases that should be prioritized if a solution is to be proposed. The most basic of which is being able to determine in the template whether a field is required or not for the sake of displaying an asterisk (or something similar).

Cheers to everybody who is willing to keep this conversation going so that, hopefully, improvements can continue to be made.

**Comment by moniuch** [comment]
For my purposes so far, being able to set a named validator - and then query for its presence - would have sufficed. 
By _named_ I mean a custom optional string passed to `.setValidator` (and `Validators.compose`, etc). Something like: 

```
this.mycontrol.setValidators(Validators.required, 'required');
```

Then, of course, I should be able to override a named validator by passing a new function (or noop/void/null) with the identical name.
And obviously, I should be able to query an AbstractControl later whether it has a given named validator.

@lazarljubenovic writes:
> It's a tricky one. While I understand that it's the quickest and most obvious solution to simply “get” the validator, it won't solve much. You'll add a validator Validate.range(1, 5), and you'll “get” it (“it” being the resulting function), but you won't be able to “get” the numbers 1 and 5. 

Why would one need to retrieve args? If I have set it earlier, then I know the params. When setting `Validate.range(1,5)` on a control, I could set it as named validator under a parsable name, say,

```
this.mycontrol.setValidators(Validate.range(1,5), 'validate:range:1:5')
```

and I would be a happy camper. Should I need to change it - I'd pass null or noop to override it:

```
this.mycontrol.setValidators(null, 'validate:range:1:5')
```

and it would be gone from the control.

I really don't need over-complicated solutions.

In the most prominent case that lots of developers seem to be after in this ticket, it's about being able to label a `required` validator, and have a directive that can automatically place an asterisk next to that field. With named validators you'd be also able to style the label or even set the validation error message or a help hint.

So in general I disagree in that _it won't solve much_ - it would solve a whole lot for a lot of apps! 

Maybe one could achieve that with a custom directive, IDK. All in all, IMO the current ReactiveForms API doesn't really deserve to be maintained, it is too imperative, and weakly typed too. Angular needs something more declarative with conditional validation, therefore I won't be hoping for anything from a 4 year old ticket ;-)

Have a great weekend everyone :)

**Comment by lazarljubenovic** [comment]
> Why would one need to retrieve args? If I have set it earlier, then I know the params. 

By that logic, why do you need to retrieve the validator at all? If you set it earlier, then you know that you set it.

**Comment by patrykdawid** [comment]
Why did you only comment on a small piece of the entire substantive message?

This way we will not come to a solution

**Comment by moniuch** [comment]
> By that logic, why do you need to retrieve the validator at all? If you set it earlier, then you know that you set it.

Because it's usually two different parties involved here, with distinct responsibilities:
- The container is responsible for setting validators and managing the entire form.
- A directive on a specific control in the template would be solely responsible for looking into that control's validators and do its presentational job.

Therefore the latter should be able to use a getter.

Thanks to all for the discussion.

**Comment by mlc-mlapis** [comment]
Mainly related to:
https://github.com/angular/angular/pull/37263#issuecomment-723333060
https://github.com/angular/angular/pull/39235

**Comment by khodabandelu** [comment]
+1

**Comment by dylhunn** [comment]
**Proposal and request for comment**: below is a summary of the current state of this issue, and a proposed alternative solution.

## Problem Summary

Currently, all form controls that extend the `AbstractControl` class (such as `FormArray`, `FormGroup`, `FormControl`) have the `setValidators` method. This method overrides the entire set of validators, with no easy way to add or remove a particular validator function once the form control is created. This could be fixed by adding a new API on `AbstractControl`.

## Why is This Needed?

After reviewing this thread, it seems there are three main uses cases for this feature:

1. Add validators after form control creation, as a reaction to a user action.
**Example**: make an `<input [formControl]=”control”>` required (by adding `Validators.required` to the underlying `FormControl` instance) if some checkbox is checked in the UI. 


2. Remove validators after form control creation, as a reaction to a user action.
**Example**: make an `<input [formControl]=”control”>` optional by removing the `Validators.required` validator.


3. Check whether a control has a particular validator applied.
**Example**: display a different UI based on whether a validator is present, as below (hypothetical code).

    ```
    <input type="text" [formControl]="control">
    <span *ngIf="control.hasValidator(Validators.required)">* required</span>
    ```
    (This would be particularly useful with dynamic forms.)

## A plain `getValidators` & `setValidators` API is problematic

Since each `AbstractControl` already retains its own validator set, we could simply expose it via a `getValidators` method. However, there are some issues with this approach:

1. `AbstractControl.setValidators` is a destructive operation that might bring controls and directives out-of-sync. We may want to discourage `setValidators` for these use cases, in favor of `addValidators`/`removeValidators` calls.
2. Form controls might be consumed by different components or directives, and `setValidators` is a blunt instrument that might accidentally override validators added by others.
3. Adding `getValidators` would require extending API support to also include `Validator`/`AsyncValidator` classes. These interfaces contain functions that will not be triggered when validators are just set via `setValidators` call.
4. A plain `getValidators` API will return unnamed function objects, which could be confusing -- there is very little one can do with an array of anonymous functions. Some of the comments on this issue propose adding another parameter to `setValidators` so each validator is associated with a name. We considered this solution, but it would require a redesign of the Forms API, to ensure names are provided properly at all of the many places validators can be attached.

## The Alternative Proposal

As an alternative to providing direct access to the list of validators, the following API would allow callers to add and remove validators individually:

```
AbstractControl.addValidators(validators: ValidatorFn|ValidatorFn[]): void;
AbstractControl.removeValidators(validators: ValidatorFn|ValidatorFn[]): void;
AbstractControl.hasValidator(validator: ValidatorFn): boolean;

AbstractControl.addAsyncValidators(validators: AsyncValidatorFn|AsyncValidatorFn[]): void;
AbstractControl.removeAsyncValidators(validators: AsyncValidatorFn|AsyncValidatorFn[]): void;
AbstractControl.hasAsyncValidator(validator: AsyncValidatorFn): boolean;
```

This API appears to address the above use cases at least as well as `getValidators`, while avoiding some of the above problems.

## Example

This API operates on *function references*. To illustrate that fact, below is a simple example.

First, we create a form control and a validator function, which is not present at first:
```
const formControl = new FormControl(0);
const minValidator = Validators.min(10);
formControl.hasValidator(minValidator); // false
```

Then, we add it, and check its presence using the function reference:
```
formControl.addValidator(minValidator);
formControl.hasValidator(minValidator); // true
```

Finally, we remove and check it by reference once more:
```
formControl.removeValidator(minValidator);
formControl.hasValidator(minValidator); // false
```

## Appendix: API doesn't work with directives

This is a major caveat, and one of the principal reasons this useful feature wasn't implemented sooner.

Imagine that your template declares a validator directly in the HTML:
```
<input type=”number” [formControl]=”myFormControl” required>
```

This uses a directive (which uses Validators.required internally) instead of a function directly, so it is impossible for caller code to have a reference to the validator function. To be specific, the following will not work:
```
formControl.hasValidator(Validators.required); // False, because this is a different function reference!
```

Although `required` is used in this example, this caveat applies to *any* validator provided as a directive.

---

We are interested in comments about this proposal: both the API design, and whether it would solve for your use case. Please feel free to comment below.

**Comment by PowerKiKi** [comment]
I hope I'm wrong, but I don't think this proposal would allow us to solve https://github.com/angular/components/issues/2574, according to the mentioned caveat.

This is because if my own code set `Validators.required` in a form control, then `@angular/material` has no way to know which function reference to use to check for the required validator presence. Is this correct ?

I was under the impression that this use-case was one of the major thing to solve.

If associating a validator with a name would properly solve all use-cases, maybe a redesign of the Forms API, that also include all the good things from #31963, should be considered, instead of doing half-measures ? A sort of `ReactiveFormsV2Module` ?

**Comment by dylhunn** [comment]
@PowerKiKi If `Validators.required` is set from TypeScript, either from `setValidators` or a `FormControl` constructor, then it should be the same function reference. It looks like your example uses `new FormControl('hello', Validators.required)`, so `hasValidator(Validators.required)` would be `true`. It's only when the validator comes from a directive that it gets wrapped in a problematic way. Hopefully I read your linked issue correctly!

**Comment by PowerKiKi** [comment]
I was not aware that the directive wraps the validator in a problematic way. So it's a relief that a "normal usage" of `Validators.required` would indeed be detectable by `@angular/material`. That's the most frequent and annoying thing from my point of view. And I'm looking forward to seeing it fixed.

However this would not be true for any other validators that takes an argument, such as `Validators.min`. Similar to the required validator that should automatically show a `*` to the user, I would like that the `Validators.min(3)` set by TypeScript is reflected as the attribute `min="3"` in DOM, so that browsers can behave accordingly (prevent input of invalid things). Unfortunately it would seems that I would have to wait for that future Forms redesign.

**Comment by lazarljubenovic** [comment]
> make an `<input [formControl]=”control”>` required (by adding Validators.required to the underlying FormControl instance) if some checkbox is checked in the UI.

Put a validator on a group instead, which validates the combination of the checkbox and the input field. This is a more declarative and pure approach which is easier to debug and reason about. This aligns with the general mantra that most frameworks have: describe _what_ should happen, not _how_ it should happen. Do not manually toggle presence of a class in the view; write a declarative template instead. Manually toggling a validator based on another value is exactly what Angular is built to escape from. Angular should not encourage writing jQuery-like code.

> make an `<input [formControl]=”control”>` optional by removing the Validators.required validator.

Same as above.

> display a different UI based on whether a validator is present

I agree with this. Validators should be queryable, for them to be useful they must not be just functions. A validator should be a class which implements an interface. This will then allow me to include meta-data, which I can use to write the error message, or a hint, or whatever. Imagine this:

```ts
export class RequiredValidator implements Validator {

  public isValid (value: any) { return value != null && value != '' }

  public readonly errorMessage = `This field is required.` // now I can read this in template

}
```

**Comment by dylhunn** [comment]
> I agree with this. Validators should be queryable, for them to be useful they must not be just functions. A validator should be a class which implements an interface. This will then allow me to include meta-data, which I can use to write the error message, or a hint, or whatever.

@lazarljubenovic I agree. Validators as bare functions, much like event handlers (which are also raw functions), leave a lot to be desired when removing or querying them at a later time.

This would be a pretty major change though, which might fit better with a potentially upcoming revamp of the forms module. Our general thinking is that, once we complete typed forms (hopefully in v13), it will be easier to make major changes with accompanying schematics, and we can do a revamp.

We were hoping to address some of the key uses (namely simple stuff like checking for `required`) without having to wait on major changes, but we will discuss further.

**Comment by dylhunn** [comment]
Hi everyone! #42838 has now been merged, and should become available in v12.2.

**Comment by franckyin** [comment]
I am really a noob at all this but by having reached this topic through a google search 30 minutes ago and after scrolling the thread full of hope, and seeing so many people having the same needs then @dylhunn making a commit one hour ago feels amazing. Thank you @dylhunn I cannot wait to play with the feature!

**Comment by agborkowski** [comment]
@dylhunn 🍻
